src/mame/tvgames/xavix.h

```diff
@@ -854,25 +921,51 @@ class xavix_cart_state : public xavix_state
 
 protected:
 
+	virtual void xavix_extbus_map(address_map &map) override ATTR_COLD;
+
+	u8 cart_r(offs_t offset)
+	{
+		if (m_cartslot->has_cart())
+		{
+			return m_cartslot->read_cart(offset);
+		}
+		else
+		{
+			return m_rgn[(offset) & (m_rgnlen - 1)];
+		}
+	}
+
+	void cart_w(offs_t offset, uint8_t data)
+	{
+		if (m_cartslot->has_cart())
+		{
+			m_cartslot->write_cart(offset, data);
+		}
+		else
+		{
+			logerror("%s: unhandled write access to cart area with no cart installed %08x %02x\n", machine().describe_context(), offset, data);
+		}
+	}
+
 	// for Cart cases this memory bypass becomes more complex
 
 	virtual uint8_t opcodes_000000_r(offs_t offset) override
 	{
 		if (offset & 0x8000)
 		{
-			if ((offset & 0x7fffff) >= m_cartlimit)
+			if (m_disable_memory_bypass)
 			{
-				return m_rgn[(offset) & (m_rgnlen - 1)];
+				return m_maincpu->space(6).read_byte(offset & 0x7fffff);
 			}
 			else
 			{
-				if (m_cartslot->has_cart())
+				if ((offset & 0x7fffff) >= m_cartlimit)
 				{
-					return m_cartslot->read_cart(offset);
+					return m_rgn[(offset) & (m_rgnlen - 1)];
 				}
 				else
 				{
-					return m_rgn[(offset) & (m_rgnlen - 1)];
+					return cart_r(offset);
 				}
 			}
 		}
```

Comment:
This is another place you’re hurting performance by doing a pile of checks on every access that should be possible to avoid entirely by installing things in the right place on start (`offset & 0x8000`, `(offset & 0x7fffff) >= m_cartlimit`, etc.) or using a memory view (`m_disable_memory_bypass`).

Reply:
likewise, that's what is planned for a future PR, not this one